include/xgboost/base.h

```diff
@@ -259,10 +259,61 @@ class GradientPairInternal {
 using GradientPair = detail::GradientPairInternal<float>;
 /*! \brief High precision gradient statistics pair */
 using GradientPairPrecise = detail::GradientPairInternal<double>;
-/*! \brief Fixed point representation for gradient pair. */
-using GradientPairInt32 = detail::GradientPairInternal<int>;
-/*! \brief Fixed point representation for high precision gradient pair. */
-using GradientPairInt64 = detail::GradientPairInternal<int64_t>;
+
+/*! \brief Fixed point representation for high precision gradient pair. Has a different interface so
+ * we don't accidentally use it in gain calculations.*/
+class GradientPairInt64 {
```

Comment:
I think this duplicated can be removed with some type checking in the `XGBOOST_DEVICE explicit GradientPairInternal(const GradientPairInternal<T2> &g)` constructor and similar functions.

In c++ 17:
``` c++
template <typename... T>
struct is_grad_convertible {
  using type = std::conditional_t<(std::is_floating_point<T>::value && ...) ||
                                      (std::is_integral<T>::value && ...),
                                  std::true_type, std::false_type>;
};

// in function templates:
template <typename T2, std::enable_if_t<is_grad_convertible<T, T2>::value>* = nullptr>
GradientPairInternal(const GradientPairInternal<T2> &g);
```

Reply:
It's not the copying that's the problem. When I began to change the gradient types the following statements still compile with integer gradients and they really shouldn't:
`if(grad.GetHess() >= min_child_weight)`
`evaluator.CalcGain(...., grad,...)`
So by changing the interface name these statements for longer compile and it is not possible to accidentally use the quantised value in calculations.

It's a bit annoying to duplicate some code but I think it's worth it. I kept finding bugs from the above situation until I made this change.